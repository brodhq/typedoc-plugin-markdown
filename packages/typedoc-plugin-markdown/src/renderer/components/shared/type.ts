import {
  DeclarationReflection,
  ParameterReflection,
  SignatureReflection,
} from 'typedoc';
import {
  ArrayType,
  ConditionalType,
  IndexedAccessType,
  InferredType,
  IntersectionType,
  IntrinsicType,
  LiteralType,
  PredicateType,
  QueryType,
  ReferenceType,
  ReflectionType,
  TupleType,
  Type,
  TypeOperatorType,
  TypeParameterType,
  UnionType,
  UnknownType,
} from 'typedoc/dist/lib/models/types';

import { escapeChars } from '../../utils';
import { LinkComponent } from './link';

type Collapse = 'object' | 'function' | 'all' | 'none';
type allowedTypes =
  | ArrayType
  | IntersectionType
  | IntrinsicType
  | ReferenceType
  | TupleType
  | UnionType
  | TypeOperatorType
  | TypeParameterType
  | QueryType
  | PredicateType
  | ReferenceType
  | ConditionalType
  | IndexedAccessType
  | UnknownType
  | InferredType
  | Type
  | ReflectionType
  | DeclarationReflection
  | ParameterReflection;

export function TypeComponent(
  type: allowedTypes,
  collapse: Collapse = 'none',
  emphasis = true,
) {
  if (type instanceof ReferenceType) {
    return getReferenceType(type, emphasis);
  }

  if (type instanceof ArrayType && type.elementType) {
    return getArrayType(type, emphasis);
  }

  if (type instanceof UnionType && type.types) {
    return getUnionType(type, emphasis);
  }

  if (type instanceof IntersectionType && type.types) {
    return getIntersectionType(type);
  }

  if (type instanceof TupleType && type.elements) {
    return getTupleType(type);
  }

  if (type instanceof IntrinsicType && type.name) {
    return getIntrinsicType(type, emphasis);
  }

  if (type instanceof ReflectionType && type.declaration) {
    return getReflectionType(type.declaration, collapse);
  }

  if (type instanceof DeclarationReflection) {
    return getReflectionType(type, collapse);
  }

  if (type instanceof TypeOperatorType) {
    return getTypeOperatorType(type);
  }

  if (type instanceof QueryType) {
    return getQueryType(type);
  }

  if (type instanceof TypeParameterType) {
    return getTypeParameterType(type);
  }

  if (type instanceof ConditionalType) {
    return getConditionalType(type);
  }

  if (type instanceof IndexedAccessType) {
    return getIndexAccessType(type);
  }

  if (type instanceof UnknownType) {
    return getUnknownType(type);
  }

  if (type instanceof InferredType) {
    return getInferredType(type);
  }

  if (type instanceof LiteralType) {
    return getLiteralType(type);
  }

  return type ? escapeChars(type.toString()) : '';
}

function getLiteralType(model: LiteralType) {
  if (typeof model.value === 'bigint') {
    return `\`${model.value}n\``;
  }
  return `\`\`${JSON.stringify(model.value)}\`\``;
}

function getReflectionType(model: DeclarationReflection, collapse: Collapse) {
  if (model.signatures) {
    return collapse === 'function' || collapse === 'all'
      ? `\`fn\``
      : getFunctionType(model.signatures);
  }
  return collapse === 'object' || collapse === 'all'
    ? `\`Object\``
    : getDeclarationType(model);
}

function getDeclarationType(model: DeclarationReflection) {
  if (model.indexSignature || model.children) {
    let indexSignature = '';
    const declarationIndexSignature = model.indexSignature;
    if (declarationIndexSignature) {
      const key = declarationIndexSignature.parameters
        ? declarationIndexSignature.parameters.map(
            (param) => `[${param.name}: ${param.type}]`,
          )
        : '';

      const obj = declarationIndexSignature.type
        ? TypeComponent(declarationIndexSignature.type as allowedTypes)
        : '`unknown`';
      indexSignature = `${key}: ${obj}; `;
    }
    const types =
      model.children &&
      model.children.map((obj) => {
        return `\`${obj.name}${
          obj.flags.isOptional ? '?' : ''
        }\`: ${TypeComponent(
          obj.signatures || obj.children ? obj : (obj.type as allowedTypes),
        )} ${
          obj.defaultValue && obj.defaultValue !== '...'
            ? `= ${escapeChars(obj.defaultValue)}`
            : ''
        }`;
      });
    return `{ ${indexSignature ? indexSignature : ''}${
      types ? types.join('; ') : ''
    } }${
      model.defaultValue && model.defaultValue !== '...'
        ? `= ${escapeChars(model.defaultValue)}`
        : ''
    }`;
  }
  return '{}';
}

export function getFunctionType(modelSignatures: SignatureReflection[]) {
  const functions = modelSignatures.map((fn) => {
    const typeParams = fn.typeParameters
      ? `<${fn.typeParameters
          .map((typeParameter) => typeParameter.name)
          .join(', ')}\\>`
      : [];
    const params = fn.parameters
      ? fn.parameters.map((param) => {
          return `${param.flags.isRest ? '...' : ''}\`${param.name}${
            param.flags.isOptional ? '?' : ''
          }\`: ${TypeComponent(
            param.type ? param.type : (param as allowedTypes),
          )}`;
        })
      : [];
    const returns = fn.type ? TypeComponent(fn.type) : '`unknown`';
    return typeParams + `(${params.join(', ')}) => ${returns}`;
  });
  return functions.join('');
}

function getReferenceType(model: ReferenceType, emphasis) {
  if (model.reflection || (model.name && model.typeArguments)) {
    const reflection =
      model.reflection && model.reflection.url
        ? [
            `${LinkComponent(
              `\`${model.reflection.name}\``,
              model.reflection.url,
              false,
            )}`,
          ]
        : [`\`${model.name}\``];
    if (model.typeArguments && model.typeArguments.length > 0) {
      reflection.push(
        `<${model.typeArguments
          .map((typeArgument) => TypeComponent(typeArgument, 'all'))
          .join(', ')}\\>`,
      );
    }
    return reflection.join('');
  }
  return emphasis ? `\`${model.name}\`` : escapeChars(model.name);
}

function getArrayType(model: ArrayType, emphasis: boolean) {
  const arrayType = TypeComponent(model.elementType, 'none', emphasis);
  return model.elementType.type === 'union'
    ? `(${arrayType})[]`
    : `${arrayType}[]`;
}

function getUnionType(model: UnionType, emphasis: boolean) {
  return model.types
    .map((unionType) => TypeComponent(unionType, 'none', emphasis))
    .join(` \\| `);
}

function getIntersectionType(model: IntersectionType) {
  return model.types
    .map((intersectionType) => TypeComponent(intersectionType))
    .join(' & ');
}

function getTupleType(model: TupleType) {
  return `[${model.elements
    .map((element) => TypeComponent(element))
    .join(', ')}]`;
}

function getIntrinsicType(model: IntrinsicType, emphasis: boolean) {
  return emphasis ? `\`${escapeChars(model.name)}\`` : escapeChars(model.name);
}

function getTypeOperatorType(model: TypeOperatorType) {
  return `${model.operator} ${TypeComponent(model.target)}`;
}

function getQueryType(model: QueryType) {
  return `typeof ${TypeComponent(model.queryType)}`;
}

function getTypeParameterType(model: TypeParameterType) {
  return escapeChars(model.name);
}

function getInferredType(model: InferredType) {
  return `infer ${escapeChars(model.name)}`;
}

function getUnknownType(model: UnknownType) {
  return escapeChars(model.name);
}

function getConditionalType(model: ConditionalType) {
  const md: string[] = [];
  if (model.checkType) {
    md.push(TypeComponent(model.checkType));
  }
  md.push('extends');
  if (model.extendsType) {
    md.push(TypeComponent(model.extendsType));
  }
  md.push('?');
  if (model.trueType) {
    md.push(TypeComponent(model.trueType));
  }
  md.push(':');
  if (model.falseType) {
    md.push(TypeComponent(model.falseType));
  }
  return md.join(' ');
}

function getIndexAccessType(model: IndexedAccessType) {
  const md: string[] = [];
  if (model.objectType) {
    md.push(TypeComponent(model.objectType));
  }
  if (model.indexType) {
    md.push(`[${TypeComponent(model.indexType)}]`);
  }
  return md.join('');
}
