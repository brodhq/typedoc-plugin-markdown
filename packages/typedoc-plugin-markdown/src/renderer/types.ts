import {
  DeclarationReflection,
  ProjectReflection,
  TypeDocOptionMap,
} from 'typedoc/dist';

export interface State {
  project: ProjectReflection;
  projectName: string;
  projectUrl: string;
  currentUrl: string;
  currentModel: DeclarationReflection | null;
  options: Partial<PluginOptions>;
  hasReadme: boolean | null;
  globalsName: string;
}

export interface PluginOptions extends TypeDocOptionMap {
  entryDocument: string;
  hideBreadcrumbs: boolean;
  publicPath: string;
  hideInPageTOC: boolean;
  indexTitle: string;
  namedAnchors: boolean;
}
