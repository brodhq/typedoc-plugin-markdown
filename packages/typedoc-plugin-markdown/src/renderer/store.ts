import { ProjectReflection } from 'typedoc';

import { PluginOptions, State } from './types';

const state: State = {
  project: new ProjectReflection(''),
  currentModel: null,
  hasReadme: null,
  currentUrl: '',
  projectName: '',
  globalsName: '',
  projectUrl: '',
  options: {},
};

export const useState = () => state;

export function setDefaultState(
  project: ProjectReflection,
  options: Partial<PluginOptions>,
) {
  setState({
    projectName: project.name,
    projectUrl: project.url,
    project: project,
    globalsName:
      options.entryPoints && options.entryPoints.length > 1
        ? 'Modules'
        : 'Exports',
    options: options,
    hasReadme: options.readme !== 'none',
  });
}

export function setState(nextState) {
  Object.assign(state, nextState);
}
