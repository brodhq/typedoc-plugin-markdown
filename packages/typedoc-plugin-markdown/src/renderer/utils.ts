export function escapeChars(str: string) {
  return str
    .replace(/>/g, '\\>')
    .replace(/_/g, '\\_')
    .replace(/`/g, '\\`')
    .replace(/\|/g, '\\|');
}

export function stripLineBreaks(str: string) {
  return str
    ? str.replace(/\n/g, ' ').replace(/\r/g, ' ').replace(/\t/g, ' ').trim()
    : '';
}

export function stripComments(str: string) {
  return str
    .replace(/(?:\/\*(?:[\s\S]*?)\*\/)|(?:^\s*\/\/(?:.*)$)/g, ' ')
    .replace(/\n/g, '')
    .replace(/^\s+|\s+$|(\s)+/g, '$1');
}

export function spaces(length: number) {
  return `!spaces${[...Array(length)].map(() => ' ').join('')}`;
}
