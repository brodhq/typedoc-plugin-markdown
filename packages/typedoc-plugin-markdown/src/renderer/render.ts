import * as path from 'path';

import ProgressBar from 'progress';
import {
  DeclarationReflection,
  ProjectReflection,
  Reflection,
  ReflectionKind,
  UrlMapping,
} from 'typedoc';
import { RendererEvent } from 'typedoc/dist/lib/output/events';
import { TemplateMapping } from 'typedoc/dist/lib/output/themes/DefaultTheme';

import { setDefaultState, setState } from './store';
import { PluginOptions } from './types';

export async function render(
  project: ProjectReflection,
  outputDirectory: string,
) {
  if (
    !this.prepareTheme() ||
    !(await this.prepareOutputDirectory(outputDirectory))
  ) {
    return;
  }

  const output = new RendererEvent(
    RendererEvent.BEGIN,
    outputDirectory,
    project,
  );

  const options =
    this.application.options.getRawValues() as unknown as Partial<PluginOptions>;

  output.settings = options;
  output.urls = getUrls(project, options);
  if (output.urls) {
    setDefaultState(project, options);

    if (output.urls) {
      const bar = new ProgressBar('Rendering [:bar] :percent', {
        total: output.urls.length,
        width: 40,
      });
      this.trigger(output);
      if (!output.isDefaultPrevented) {
        output.urls?.forEach((mapping: UrlMapping, i) => {
          const page = output.createPageEvent(mapping);
          setState({ currentModel: page.model, currentUrl: page.url });
          this.renderDocument(page);
          bar.tick();
        });
        this.trigger(RendererEvent.END, output);
      }

      setState(null);
    }
  }
}

export function getUrls(
  project: ProjectReflection,
  options: any,
): UrlMapping[] {
  const urls: UrlMapping[] = [];
  const noReadmeFile = options.readme == path.join(process.cwd(), 'none');
  if (noReadmeFile) {
    project.url = options.entryDocument;
    urls.push(new UrlMapping(options.entryDocument, project, 'reflection.hbs'));
  } else {
    project.url = 'modules.md';
    urls.push(new UrlMapping('modules.md', project, 'reflection.hbs'));
    urls.push(new UrlMapping(options.entryDocument, project, 'readme.hbs'));
  }
  project.children?.forEach((child: Reflection) => {
    if (child instanceof DeclarationReflection) {
      buildUrls(child as DeclarationReflection, urls);
    }
  });
  return urls;
}

function buildUrls(
  reflection: DeclarationReflection,
  urls: UrlMapping[],
): UrlMapping[] {
  const URL_PREFIX = /^(http|ftp)s?:\/\//;
  const mapping = mappings.find((mapping) => reflection.kindOf(mapping.kind));

  if (mapping) {
    if (!reflection.url || !URL_PREFIX.test(reflection.url)) {
      const url = toUrl(mapping, reflection);
      urls.push(new UrlMapping(url, reflection, mapping.template));
      reflection.url = url;
      reflection.hasOwnDocument = true;
    }
    for (const child of reflection.children || []) {
      if (mapping.isLeaf) {
        applyAnchorUrl(child, reflection);
      } else {
        buildUrls(child, urls);
      }
    }
  } else if (reflection.parent) {
    applyAnchorUrl(reflection, reflection.parent);
  }
  return urls;
}

function toUrl(mapping: TemplateMapping, reflection: DeclarationReflection) {
  return mapping.directory + '/' + getUrl(reflection) + '.md';
}

function getUrl(reflection: Reflection, relative?: Reflection): string {
  let url = reflection.getAlias();

  if (
    reflection.parent &&
    reflection.parent !== relative &&
    !(reflection.parent instanceof ProjectReflection)
  ) {
    url = getUrl(reflection.parent, relative) + '.' + url;
  }

  return url;
}

function applyAnchorUrl(reflection: Reflection, container: Reflection) {
  const URL_PREFIX = /^(http|ftp)s?:\/\//;
  if (!reflection.url || !URL_PREFIX.test(reflection.url)) {
    const reflectionId = reflection.name.toLowerCase();
    const anchor = toAnchorRef(reflectionId);
    reflection.url = container.url + '#' + anchor;
    reflection.anchor = anchor;
    reflection.hasOwnDocument = false;
  }
  reflection.traverse((child) => {
    if (child instanceof DeclarationReflection) {
      applyAnchorUrl(child, container);
    }
  });
}

function toAnchorRef(reflectionId: string) {
  return reflectionId;
}

const mappings = [
  {
    kind: [ReflectionKind.Module],
    isLeaf: false,
    directory: 'modules',
    template: 'reflection.hbs',
  },
  {
    kind: [ReflectionKind.Namespace],
    isLeaf: false,
    directory: 'modules',
    template: 'reflection.hbs',
  },
  {
    kind: [ReflectionKind.Enum],
    isLeaf: false,
    directory: 'enums',
    template: 'reflection.hbs',
  },
  {
    kind: [ReflectionKind.Class],
    isLeaf: false,
    directory: 'classes',
    template: 'reflection.hbs',
  },
  {
    kind: [ReflectionKind.Interface],
    isLeaf: false,
    directory: 'interfaces',
    template: 'reflection.hbs',
  },
];
