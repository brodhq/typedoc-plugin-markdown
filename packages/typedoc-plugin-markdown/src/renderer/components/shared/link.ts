import * as path from 'path';

import { link } from '../../../markdown-tools/elements';
import { useState } from '../../store';
import { escapeChars } from '../../utils';

export function LinkComponent(label: string, to = '', escapeLabel = true) {
  label = escapeLabel ? escapeChars(label) : label;
  return link(label, getUrl(to));
}

function getUrl(to: string) {
  const { options, currentUrl } = useState();

  const urlPrefix = /^(http|ftp)s?:\/\//;
  if (urlPrefix.test(to)) {
    return to;
  } else {
    return options.publicPath
      ? getPublicUrl(options.publicPath, to)
      : getRelativeUrl(currentUrl, to);
  }
}

function getPublicUrl(publicPath: string, url: string) {
  return publicPath + url;
}

function getRelativeUrl(from: string, to: string) {
  const relative = path.relative(path.dirname(from), path.dirname(to));
  return path.join(relative, path.basename(to)).replace(/\\/g, '/');
}
