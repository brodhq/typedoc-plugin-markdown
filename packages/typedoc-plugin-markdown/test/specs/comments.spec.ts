import { CommentsComponent } from '../../src/renderer/components/shared/comments';
import { setState } from '../../src/renderer/store';
import { TestApp } from '../test-app';

describe(`Comments:`, () => {
  let testApp: TestApp;

  beforeAll(async () => {
    testApp = new TestApp(['comments.ts', 'declarations.ts']);
    await testApp.bootstrap();
  });

  test(`should build @link reference'`, () => {
    const page = testApp.getPageByModelName('comments').model;
    setState({ currentModel: page });
    const reflection = testApp.project.findReflectionByName(
      'commentsWithDocLinks',
    );
    expect(CommentsComponent(reflection.comment)).toMatchSnapshot();
  });

  test(`should convert symbols brackets to symbol links'`, () => {
    const page = testApp.getPageByModelName('comments').model;
    setState({ currentModel: page });
    const reflection = testApp.project.findReflectionByName(
      'commentsWithSymbolLinks',
    );
    expect(CommentsComponent(reflection.comment)).toMatchSnapshot();
  });

  test(`should convert comments with tags'`, () => {
    const page = testApp.getPageByModelName('comments').model;
    setState({ currentModel: page });
    const reflection = testApp.project.findReflectionByName('commentsWithTags');
    expect(CommentsComponent(reflection.comment)).toMatchSnapshot();
  });

  test(`should allow html in comments'`, () => {
    const page = testApp.getPageByModelName('comments').model;
    setState({ currentModel: page });
    const reflection = testApp.project.findReflectionByName('commentsWithHTML');
    expect(CommentsComponent(reflection.comment)).toMatchSnapshot();
  });
});
