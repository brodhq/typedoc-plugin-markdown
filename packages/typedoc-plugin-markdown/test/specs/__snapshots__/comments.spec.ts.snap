// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comments: should allow html in comments' 1`] = `
"<p>
You can write <strong>HTML</strong> tags directly in <a href=\\"comments\\">comments</a>
</p>
<ul>
<li>List item</li>
</ul>"
`;

exports[`Comments: should build @link reference' 1`] = `
"- [\`CommentClass\`](RelativeUrlComponent(reflection.url))
- [CommentClass.comment](RelativeUrlComponent(reflection.url))
- {@link unknownA}
- [commentsInReturn](RelativeUrlComponent(reflection.url))
- [Google](https://www.google.com)
- [GitHub](https://github.com)"
`;

exports[`Comments: should convert comments with tags' 1`] = `
"**\`name\`** Tag description on same line

**\`description\`**
Tag description on new line

- Tag description on another line

**\`deprecated\`**
Another tag description"
`;

exports[`Comments: should convert symbols brackets to symbol links' 1`] = `
"Additionally you can link to other classes, members or functions using double square brackets.

- [CommentClass](RelativeUrlComponent(reflection.url))
- [commentsInReturn](RelativeUrlComponent(reflection.url))
- [[unknownB]]"
`;
